% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math} % this also loads fontspec
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else
  % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[margin=1in]{geometry}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage{bookmark}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same}
\hypersetup{
  pdftitle={R-assignment},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}

\title{R-assignment}
\author{}
\date{\vspace{-2.5em}2025-03-19}

\begin{document}
\maketitle

\subsection{R Markdown}\label{r-markdown}

\section{README}\label{readme}

\section{Read the files and store them in
variables}\label{read-the-files-and-store-them-in-variables}

fang\_data \textless- read.table(``fang\_et\_al\_genotypes.txt'', header
= TRUE, sep = ``\t", stringsAsFactors = FALSE) snp\_data \textless-
read.table(''snp\_position.txt'', header = TRUE, sep = ``\t",
stringsAsFactors = FALSE) \# Inspect dimensions and size of the data
dim(fang\_data) \# Dimensions of the fang\_et\_al\_genotypes data
dim(snp\_data) \# Dimensions of the snp\_position data fang\_file\_size
\textless- file.size(''fang\_et\_al\_genotypes.txt'') snp\_file\_size
\textless- file.size(``snp\_position.txt'') fang\_file\_size\_kb
\textless- fang\_file\_size / 1024 snp\_file\_size\_kb \textless-
snp\_file\_size / 1024 cat(``File size of
fang\_et\_al\_genotypes.txt:'', fang\_file\_size, ``bytes ('',
fang\_file\_size\_kb, ``KB)\n'') cat(``File size of
snp\_position.txt:'', snp\_file\_size, ``bytes ('', snp\_file\_size\_kb,
``KB)\n'') \#maize data processing \# Read the files and store them in
variables fang\_data \textless-
read.table(``fang\_et\_al\_genotypes.txt'', header = TRUE, sep = ``\t",
stringsAsFactors = FALSE) snp\_data \textless-
read.table(''snp\_position.txt'', header = TRUE, sep = ``\t",
stringsAsFactors = FALSE)

\section{Transpose the fang\_et\_al\_genotypes
data}\label{transpose-the-fang_et_al_genotypes-data}

print(``Transposing fang data\ldots{}'') transposed\_genotypes
\textless- t(fang\_data) transposed\_genotypes \textless-
as.data.frame(transposed\_genotypes) write.table(transposed\_genotypes,
file = ``transposed\_genotypes.txt'', sep = ``\t", row.names = TRUE,
col.names = FALSE, quote = FALSE)

\section{Read and modify snp\_position
file}\label{read-and-modify-snp_position-file}

snp\_lines \textless- readLines(``snp\_position.txt'') header \textless-
snp\_lines{[}1{]} modified\_lines \textless- c(header, ``\,``,''``,
snp\_lines{[}-1{]})
writeLines(modified\_lines,''snp\_position\_mod.txt'')

\section{Read the transposed data}\label{read-the-transposed-data}

transposed\_genotypes \textless-
read.table(``transposed\_genotypes.txt'', header = FALSE, sep = ``\t",
stringsAsFactors = FALSE)

\section{Identify columns to keep (maize categories: ZMMLR, ZMMIL,
ZMMMR)}\label{identify-columns-to-keep-maize-categories-zmmlr-zmmil-zmmmr}

cols\_to\_keep \textless- which(transposed\_genotypes{[}3, {]} \%in\%
c(``ZMMLR'', ``ZMMIL'', ``ZMMMR'')) maize\_data \textless-
transposed\_genotypes{[}, cols\_to\_keep{]}

\section{Save the filtered maize
data}\label{save-the-filtered-maize-data}

write.table(maize\_data, file = ``maize.txt'', sep = ``\t", row.names =
FALSE, col.names = FALSE, quote = FALSE)

\section{Read files}\label{read-files}

maize\_check \textless- read.table(``maize.txt'', header = FALSE, sep =
``\t", stringsAsFactors = FALSE) snp\_position \textless-
read.table(''snp\_position\_mod.txt'', header = FALSE, sep = ``\t",
stringsAsFactors = FALSE)

\section{Extract relevant SNP
columns}\label{extract-relevant-snp-columns}

snp\_subset \textless- snp\_position{[}, c(1, 3, 4){]} maize\_data
\textless- read.table(``maize.txt'', header = FALSE, sep = ``\t",
stringsAsFactors = FALSE) maize\_data \textless- maize\_data{[}-(1:2),
{]} \# Remove first two rows

\section{Merge SNP data with maize
genotypes}\label{merge-snp-data-with-maize-genotypes}

maize\_joined \textless- cbind(snp\_subset, maize\_data)
write.table(maize\_joined, file = ``maize\_joined.txt'', sep = ``\t",
row.names = FALSE, col.names = FALSE, quote = FALSE)

\section{Load required package}\label{load-required-package}

library(dplyr)

\section{Read the joined data}\label{read-the-joined-data}

maize\_data \textless- read.table(``maize\_joined.txt'', header = TRUE,
sep = ``\t", stringsAsFactors = FALSE)

\section{Function to check if a value is
numeric}\label{function-to-check-if-a-value-is-numeric}

is\_numeric\_str \textless- function(x) \{
!is.na(suppressWarnings(as.numeric(x))) \}

\section{Function to process and save each chromosome (sorted in
increasing
order)}\label{function-to-process-and-save-each-chromosome-sorted-in-increasing-order}

process\_chromosome \textless- function(chrom\_num, data) \{
print(paste(``Processing chromosome'', chrom\_num, ``\ldots{}''))
subset\_data \textless- data{[}data{[}{[}2{]}{]} == chrom\_num, {]}\\
numeric\_positions \textless- sapply(subset\_data{[}{[}3{]}{]},
is\_numeric\_str)
subset\_data\(SNP_numeric <- as.numeric(subset_data[[3]])
  subset_data <- subset_data[order(subset_data\)SNP\_numeric, na.last =
TRUE), {]} subset\_data{[}{[}3{]}{]}{[}!numeric\_positions{]} \textless-
``?'' subset\_data\$SNP\_numeric \textless- NULL\\
output\_file \textless- paste0(``maize\_sortedv1\_'', chrom\_num,
``.txt'') write.table(subset\_data, output\_file, sep = ``\t", row.names
= FALSE, quote = FALSE) print(paste(''File successfully written:``,
output\_file)) \}

\section{Apply function to chromosomes 1 to
10}\label{apply-function-to-chromosomes-1-to-10}

print(``Processing chromosomes 1 to 10 (increasing order)\ldots{}'')
lapply(1:10, process\_chromosome, data = maize\_data)

\section{Function to process and save each chromosome (sorted in
decreasing
order)}\label{function-to-process-and-save-each-chromosome-sorted-in-decreasing-order}

process\_chromosome\_desc \textless- function(chrom\_num, data) \{
print(paste(``Processing chromosome'', chrom\_num, ``(decreasing
order)\ldots{}'')) subset\_data \textless- data{[}data{[}{[}2{]}{]} ==
chrom\_num, {]}\\
numeric\_positions \textless- sapply(subset\_data{[}{[}3{]}{]},
is\_numeric\_str)
subset\_data\(SNP_numeric <- as.numeric(subset_data[[3]])
  subset_data <- subset_data[order(-subset_data\)SNP\_numeric, na.last =
TRUE), {]} subset\_data{[}{[}3{]}{]}{[}!numeric\_positions{]} \textless-
``-'' subset\_data\$SNP\_numeric \textless- NULL\\
output\_file \textless- paste0(``maize\_sortedv2\_'', chrom\_num,
``.txt'') write.table(subset\_data, output\_file, sep = ``\t", row.names
= FALSE, quote = FALSE) print(paste(''File successfully written:``,
output\_file)) \}

\section{Apply function to chromosomes 1 to
10}\label{apply-function-to-chromosomes-1-to-10-1}

print(``Processing chromosomes 1 to 10 (decreasing order)\ldots{}'')
lapply(1:10, process\_chromosome\_desc, data = maize\_data) \#repeating
the process for teosinte data \# Read the files and store them in
variables fang\_data \textless-
read.table(``fang\_et\_al\_genotypes.txt'', header = TRUE, sep = ``\t",
stringsAsFactors = FALSE) snp\_data \textless-
read.table(''snp\_position.txt'', header = TRUE, sep = ``\t",
stringsAsFactors = FALSE)

\section{Transpose the fang\_et\_al\_genotypes
data}\label{transpose-the-fang_et_al_genotypes-data-1}

print(``Transposing fang data\ldots{}'') transposed\_genotypes
\textless- t(fang\_data) transposed\_genotypes \textless-
as.data.frame(transposed\_genotypes) write.table(transposed\_genotypes,
file = ``transposed\_genotypes.txt'', sep = ``\t", row.names = TRUE,
col.names = FALSE, quote = FALSE)

\section{Read and modify snp\_position
file}\label{read-and-modify-snp_position-file-1}

snp\_lines \textless- readLines(``snp\_position.txt'') header \textless-
snp\_lines{[}1{]} modified\_lines \textless- c(header, ``\,``,''``,
snp\_lines{[}-1{]})
writeLines(modified\_lines,''snp\_position\_mod.txt'')

\section{Read the transposed data}\label{read-the-transposed-data-1}

transposed\_genotypes \textless-
read.table(``transposed\_genotypes.txt'', header = FALSE, sep = ``\t",
stringsAsFactors = FALSE)

\section{Identify columns to keep (teosinte categories: ZMPBA, ZMPIL,
ZMPJA)}\label{identify-columns-to-keep-teosinte-categories-zmpba-zmpil-zmpja}

cols\_to\_keep \textless- which(transposed\_genotypes{[}3, {]} \%in\%
c(``ZMPBA'', ``ZMPIL'', ``ZMPJA'')) teosinte\_data \textless-
transposed\_genotypes{[}, cols\_to\_keep{]}

\section{Save the filtered teosinte
data}\label{save-the-filtered-teosinte-data}

write.table(teosinte\_data, file = ``teosinte.txt'', sep = ``\t",
row.names = FALSE, col.names = FALSE, quote = FALSE)

\section{Read files}\label{read-files-1}

teosinte\_check \textless- read.table(``teosinte.txt'', header = FALSE,
sep = ``\t", stringsAsFactors = FALSE) snp\_position \textless-
read.table(''snp\_position\_mod.txt'', header = FALSE, sep = ``\t",
stringsAsFactors = FALSE)

\section{Extract relevant SNP
columns}\label{extract-relevant-snp-columns-1}

snp\_subset \textless- snp\_position{[}, c(1, 3, 4){]} teosinte\_data
\textless- read.table(``teosinte.txt'', header = FALSE, sep = ``\t",
stringsAsFactors = FALSE) teosinte\_data \textless-
teosinte\_data{[}-(1:2), {]} \# Remove first two rows

\section{Merge SNP data with teosinte
genotypes}\label{merge-snp-data-with-teosinte-genotypes}

teosinte\_joined \textless- cbind(snp\_subset, teosinte\_data)
write.table(teosinte\_joined, file = ``teosinte\_joined.txt'', sep =
``\t", row.names = FALSE, col.names = FALSE, quote = FALSE)

\section{Load required package}\label{load-required-package-1}

library(dplyr)

\section{Read the joined data}\label{read-the-joined-data-1}

teosinte\_data \textless- read.table(``teosinte\_joined.txt'', header =
TRUE, sep = ``\t", stringsAsFactors = FALSE)

\section{Function to check if a value is
numeric}\label{function-to-check-if-a-value-is-numeric-1}

is\_numeric\_str \textless- function(x) \{
!is.na(suppressWarnings(as.numeric(x))) \}

\section{Function to process and save each chromosome (sorted in
increasing
order)}\label{function-to-process-and-save-each-chromosome-sorted-in-increasing-order-1}

process\_chromosome \textless- function(chrom\_num, data) \{
print(paste(``Processing chromosome'', chrom\_num, ``\ldots{}''))
subset\_data \textless- data{[}data{[}{[}2{]}{]} == chrom\_num, {]}\\
numeric\_positions \textless- sapply(subset\_data{[}{[}3{]}{]},
is\_numeric\_str)
subset\_data\(SNP_numeric <- as.numeric(subset_data[[3]])
  subset_data <- subset_data[order(subset_data\)SNP\_numeric, na.last =
TRUE), {]} subset\_data{[}{[}3{]}{]}{[}!numeric\_positions{]} \textless-
``?'' subset\_data\$SNP\_numeric \textless- NULL\\
output\_file \textless- paste0(``teosinte\_sortedv1\_'', chrom\_num,
``.txt'') write.table(subset\_data, output\_file, sep = ``\t", row.names
= FALSE, quote = FALSE) print(paste(''File successfully written:``,
output\_file)) \}

\section{Apply function to chromosomes 1 to
10}\label{apply-function-to-chromosomes-1-to-10-2}

print(``Processing chromosomes 1 to 10 (increasing order)\ldots{}'')
lapply(1:10, process\_chromosome, data = teosinte\_data)

\section{Function to process and save each chromosome (sorted in
decreasing
order)}\label{function-to-process-and-save-each-chromosome-sorted-in-decreasing-order-1}

process\_chromosome\_desc \textless- function(chrom\_num, data) \{
print(paste(``Processing chromosome'', chrom\_num, ``(decreasing
order)\ldots{}'')) subset\_data \textless- data{[}data{[}{[}2{]}{]} ==
chrom\_num, {]}\\
numeric\_positions \textless- sapply(subset\_data{[}{[}3{]}{]},
is\_numeric\_str)
subset\_data\(SNP_numeric <- as.numeric(subset_data[[3]])
  subset_data <- subset_data[order(-subset_data\)SNP\_numeric, na.last =
TRUE), {]} subset\_data{[}{[}3{]}{]}{[}!numeric\_positions{]} \textless-
``-'' subset\_data\$SNP\_numeric \textless- NULL\\
output\_file \textless- paste0(``teosinte\_sortedv2\_'', chrom\_num,
``.txt'') write.table(subset\_data, output\_file, sep = ``\t", row.names
= FALSE, quote = FALSE) print(paste(''File successfully written:``,
output\_file)) \}

\section{Apply function to chromosomes 1 to
10}\label{apply-function-to-chromosomes-1-to-10-3}

print(``Processing chromosomes 1 to 10 (decreasing order)\ldots{}'')
lapply(1:10, process\_chromosome\_desc, data = teosinte\_data) \# Create
directories for sorted files dir.create(``sorted\_maize\_files'',
showWarnings = FALSE) dir.create(``teosinte\_sorted\_files'',
showWarnings = FALSE)

\section{Move maize\_sortedv files to
sorted\_maize\_files}\label{move-maize_sortedv-files-to-sorted_maize_files}

maize\_files \textless- list.files(pattern = ``\^{}maize\_sortedv'')
file.copy(maize\_files, ``sorted\_maize\_files'')
file.remove(maize\_files)

\section{Move teosinte\_sortedv files to
teosinte\_sorted\_files}\label{move-teosinte_sortedv-files-to-teosinte_sorted_files}

teosinte\_files \textless- list.files(pattern =
``\^{}teosinte\_sortedv'') file.copy(teosinte\_files,
``teosinte\_sorted\_files'') file.remove(teosinte\_files)

\#visualization \# Load libraries library(ggplot2) library(dplyr)

\section{Step 1: Read the data}\label{step-1-read-the-data}

maize\_data \textless- read.table(``maize\_joined.txt'', header = FALSE,
sep = ``\t", stringsAsFactors = FALSE) maize\_data \textless-
maize\_data{[}, 2:3{]} colnames(maize\_data) \textless-
c(''Chromosome'', ``SNP\_Position'')

teosinte\_data \textless- read.table(``teosinte\_joined.txt'', header =
FALSE, sep = ``\t", stringsAsFactors = FALSE) teosinte\_data \textless-
teosinte\_data{[}, 2:3{]} colnames(teosinte\_data) \textless-
c(''Chromosome'', ``SNP\_Position'')

\section{Step 2: Filter out invalid SNP
positions}\label{step-2-filter-out-invalid-snp-positions}

filter\_data \textless- function(data) \{ data \%\textgreater\%
filter(!is.na(SNP\_Position) \& grepl(``\footnote{0-9}+\$'',
SNP\_Position)) \%\textgreater\% mutate(SNP\_Position =
as.numeric(SNP\_Position)) \}

maize\_data \textless- filter\_data(maize\_data) teosinte\_data
\textless- filter\_data(teosinte\_data)

\section{Step 3: Combine the data}\label{step-3-combine-the-data}

maize\_data\(Species <- "Maize"
teosinte_data\)Species \textless- ``Teosinte'' combined\_data \textless-
bind\_rows(maize\_data, teosinte\_data)

\section{Ensure Chromosome is a factor with the correct
order}\label{ensure-chromosome-is-a-factor-with-the-correct-order}

chromosome\_order \textless- c(1:10, ``unknown'', ``multiple'')
combined\_data\(Chromosome <- factor(combined_data\)Chromosome, levels =
chromosome\_order)

\section{Step 4: Bar plot with error
bars}\label{step-4-bar-plot-with-error-bars}

snp\_counts \textless- combined\_data \%\textgreater\%
group\_by(Chromosome, Species) \%\textgreater\% summarise(SNP\_Count =
n(), .groups = ``drop'') \%\textgreater\% mutate(SE = sqrt(SNP\_Count))
\# Standard error for error bars

bar\_plot \textless- ggplot(snp\_counts, aes(x = Chromosome, y =
SNP\_Count, fill = Species)) + geom\_bar(stat = ``identity'', position =
``dodge'', alpha = 0.8) + \# Bar plot with dodge geom\_errorbar(aes(ymin
= SNP\_Count - SE, ymax = SNP\_Count + SE), \# Error bars position =
position\_dodge(width = 0.9), width = 0.25) + scale\_fill\_manual(values
= c(``Maize'' = ``blue'', ``Teosinte'' = ``pink'')) + \# Custom colors
labs(title = ``SNP Counts per Chromosome'', x = ``Chromosome'', y =
``SNP Count'', fill = ``Species'') + theme\_minimal()

print(bar\_plot)

\section{Step 5: Dot plot for SNP
positions}\label{step-5-dot-plot-for-snp-positions}

dot\_plot \textless- ggplot(combined\_data, aes(x = Chromosome, y =
SNP\_Position, color = Species)) + geom\_point(position =
position\_jitter(width = 0.2, height = 0), alpha = 0.6) + \# Dots with
jitter scale\_color\_manual(values = c(``Maize'' = ``blue'',
``Teosinte'' = ``pink'')) + \# Custom colors labs(title = ``SNP
Positions by Chromosome'', x = ``Chromosome'', y = ``SNP Position'',
color = ``Species'') + theme\_minimal()

print(dot\_plot) \#heterozygosity and homozygosity \# Load libraries
library(ggplot2) library(dplyr) library(tidyr)

\section{Step 1: Read the data}\label{step-1-read-the-data-1}

maize\_data \textless- read.table(``maize\_joined.txt'', header = FALSE,
sep = ``\t", stringsAsFactors = FALSE) teosinte\_data \textless-
read.table(''teosinte\_joined.txt'', header = FALSE, sep = ``\t",
stringsAsFactors = FALSE)

\section{Step 2: Define a function to classify
sites}\label{step-2-define-a-function-to-classify-sites}

classify\_sites \textless- function(data) \{ \# Select columns with
nucleotide data (from column 4 to the last column) nucleotide\_data
\textless- data{[}, 4:ncol(data){]}

\begin{verbatim}
# Classify each site as homozygous, heterozygous, or missing
result <- apply(nucleotide_data, 2, function(sample) {
    # Count homozygous, heterozygous, and missing sites
    homozygous <- sum(grepl("^(A/A|C/C|G/G|T/T)$", sample))  # Count homozygous sites
    heterozygous <- sum(grepl("^(A/C|A/G|A/T|C/A|C/G|C/T|G/A|G/C|G/T|T/A|T/C|T/G)$", sample))  # Count heterozygous sites
    missing <- sum(grepl("^(\\?/\\?|\\./\\.|./.)$", sample))  # Count missing sites
    
    # Calculate total nucleotides (excluding missing data)
    total_nucleotides <- sum(homozygous, heterozygous)
    
    # Calculate proportions (normalized to total nucleotides)
    data.frame(
        Homozygous = homozygous / total_nucleotides,
        Heterozygous = heterozygous / total_nucleotides,
        Missing = missing / total_nucleotides  # Missing data normalized to total sites
    )
})

# Combine results into a data frame
result <- do.call(rbind, result)
result$Sample <- colnames(nucleotide_data)  # Add sample names
return(result)
\end{verbatim}

\}

\section{Step 3: Classify sites for maize and
teosinte}\label{step-3-classify-sites-for-maize-and-teosinte}

maize\_classified \textless- classify\_sites(maize\_data)
teosinte\_classified \textless- classify\_sites(teosinte\_data)

\section{Add a ``Species'' column}\label{add-a-species-column}

maize\_classified\(Species <- "Maize"
teosinte_classified\)Species \textless- ``Teosinte''

\section{Combine the data}\label{combine-the-data}

combined\_classified \textless- bind\_rows(maize\_classified,
teosinte\_classified)

\section{Step 4: Calculate average percentages for maize and
teosinte}\label{step-4-calculate-average-percentages-for-maize-and-teosinte}

average\_percentages \textless- combined\_classified \%\textgreater\%
group\_by(Species) \%\textgreater\% summarise( Avg\_Homozygous =
mean(Homozygous) * 100, Avg\_Heterozygous = mean(Heterozygous) * 100,
Avg\_Missing = mean(Missing) * 100, .groups = ``drop'' )

\section{Print the average
percentages}\label{print-the-average-percentages}

print(``Average percentages of homozygous, heterozygous, and missing
sites:'') print(average\_percentages)

\section{Step 5: Reshape the data for
plotting}\label{step-5-reshape-the-data-for-plotting}

combined\_long \textless- combined\_classified \%\textgreater\%
pivot\_longer(cols = c(Homozygous, Heterozygous, Missing), names\_to =
``Type'', values\_to = ``Proportion'')

\section{Step 6: Visualize the
results}\label{step-6-visualize-the-results}

plot \textless- ggplot(combined\_long, aes(x = Sample, y = Proportion,
fill = Type)) + geom\_bar(stat = ``identity'', position = ``fill'') + \#
Stacked bar plot with normalized height facet\_wrap(\textasciitilde{}
Species, scales = ``free\_x'') + \# Separate panels for maize and
teosinte scale\_fill\_manual(values = c(``Homozygous'' = ``blue'',
``Heterozygous'' = ``pink'', ``Missing'' = ``gray'')) + \# Custom colors
labs(title = ``Proportion of Homozygous, Heterozygous, and Missing
Sites'', x = ``Sample'', y = ``Proportion'', fill = ``Site Type'') +
theme\_minimal() + theme(axis.text.x = element\_text(angle = 90, hjust =
1)) \# Rotate x-axis labels

print(plot)

\#homozygous and heterozygous \# Load required packages library(dplyr)
library(ggplot2) library(tidyr)

\section{Read the genotype file}\label{read-the-genotype-file}

genotype\_data \textless- read.table(``fang\_et\_al\_genotypes.txt'',
header = TRUE, sep = ``\t", stringsAsFactors = FALSE)

\section{Split into maize and teosinte based on column
3}\label{split-into-maize-and-teosinte-based-on-column-3}

maize\_data \textless- genotype\_data{[}genotype\_data{[}{[}3{]}{]}
\%in\% c(``ZMMIL'', ``ZMMLR'', ``ZMMMR''), {]} teosinte\_data \textless-
genotype\_data{[}genotype\_data{[}{[}3{]}{]} \%in\% c(``ZMPBA'',
``ZMPIL'', ``ZMPJA''), {]}

\section{Function to calculate missing, homozygous, and heterozygous
proportions}\label{function-to-calculate-missing-homozygous-and-heterozygous-proportions}

calculate\_counts \textless- function(df, group\_name) \{ \# Select
genotype columns (4th column onward) genotype\_matrix \textless- df{[},
4:ncol(df){]}

\# Initialize result dataframe result \textless- data.frame(Missing =
integer(nrow(df)), Homozygous = integer(nrow(df)), Heterozygous =
integer(nrow(df)), Group = group\_name) \# Add a group identifier

\# Loop through each row for (i in 1:nrow(genotype\_matrix)) \{
row\_values \textless- genotype\_matrix{[}i, {]}

\begin{verbatim}
# Count missing, homozygous, and heterozygous values
missing_count <- sum(row_values == "?/?")
homozygous_count <- sum(row_values %in% c("A/A", "T/T", "G/G", "C/C"))
heterozygous_count <- sum(!(row_values %in% c("?/?", "A/A", "T/T", "G/G", "C/C")))

# Normalize counts
total <- missing_count + homozygous_count + heterozygous_count
result[i, 1:3] <- c(missing_count / total, homozygous_count / total, heterozygous_count / total)
\end{verbatim}

\}

return(result) \}

\section{Apply function to maize and teosinte
data}\label{apply-function-to-maize-and-teosinte-data}

maize\_counts \textless- calculate\_counts(maize\_data, ``Maize'')
teosinte\_counts \textless- calculate\_counts(teosinte\_data,
``Teosinte'')

\section{Combine data for plotting}\label{combine-data-for-plotting}

combined\_counts \textless- rbind(maize\_counts, teosinte\_counts)

\section{Convert data to long format for
ggplot}\label{convert-data-to-long-format-for-ggplot}

long\_data \textless- pivot\_longer(combined\_counts, cols =
c(``Missing'', ``Homozygous'', ``Heterozygous''), names\_to =
``Category'', values\_to = ``Proportion'')

\section{Generate normalized stacked bar
plot}\label{generate-normalized-stacked-bar-plot}

ggplot(long\_data, aes(x = Group, y = Proportion, fill = Category)) +
geom\_bar(stat = ``identity'', position = ``fill'') + \# Normalized
height labs(title = ``Normalized Genotype Proportions in Maize and
Teosinte'', x = ``Group'', y = ``Proportion'') +
scale\_fill\_manual(values = c(``Missing'' = ``gray'', ``Homozygous'' =
``blue'', ``Heterozygous'' = ``red'')) + theme\_minimal()

\#my own visualization \#allele frequency \# Load required packages
library(dplyr) library(ggplot2) library(tidyr)

\section{Read the genotype file}\label{read-the-genotype-file-1}

genotype\_data \textless- read.table(``fang\_et\_al\_genotypes.txt'',
header = TRUE, sep = ``\t", stringsAsFactors = FALSE)

\section{Split into maize and teosinte based on column
3}\label{split-into-maize-and-teosinte-based-on-column-3-1}

maize\_data \textless- genotype\_data{[}genotype\_data{[}{[}3{]}{]}
\%in\% c(``ZMMIL'', ``ZMMLR'', ``ZMMMR''), {]} teosinte\_data \textless-
genotype\_data{[}genotype\_data{[}{[}3{]}{]} \%in\% c(``ZMPBA'',
``ZMPIL'', ``ZMPJA''), {]}

\section{Reshape data using pivot\_longer to make it
``tidy''}\label{reshape-data-using-pivot_longer-to-make-it-tidy}

tidy\_maize \textless- maize\_data \%\textgreater\% pivot\_longer(cols =
4:ncol(maize\_data), names\_to = ``SNP'', values\_to = ``Genotype'')
\%\textgreater\% mutate(Group = ``Maize'')

tidy\_teosinte \textless- teosinte\_data \%\textgreater\%
pivot\_longer(cols = 4:ncol(teosinte\_data), names\_to = ``SNP'',
values\_to = ``Genotype'') \%\textgreater\% mutate(Group = ``Teosinte'')

\section{Combine datasets}\label{combine-datasets}

tidy\_data \textless- bind\_rows(tidy\_maize, tidy\_teosinte)

\section{Extract individual alleles (first and second base in genotype
calls)}\label{extract-individual-alleles-first-and-second-base-in-genotype-calls}

tidy\_data \textless- tidy\_data \%\textgreater\% filter(Genotype !=
``?/?'') \%\textgreater\% \# Remove missing data mutate(Allele1 =
substr(Genotype, 1, 1), Allele2 = substr(Genotype, 3, 3))

\section{Gather alleles into a single column (normalize
format)}\label{gather-alleles-into-a-single-column-normalize-format}

tidy\_alleles \textless- tidy\_data \%\textgreater\% select(Group, SNP,
Allele1, Allele2) \%\textgreater\% pivot\_longer(cols = c(``Allele1'',
``Allele2''), names\_to = ``Allele\_Position'', values\_to = ``Allele'')

\section{Compute allele frequencies}\label{compute-allele-frequencies}

allele\_frequencies \textless- tidy\_alleles \%\textgreater\%
group\_by(Group, Allele) \%\textgreater\% summarise(Frequency = n(),
.groups = ``drop'') \%\textgreater\% mutate(Proportion = Frequency /
sum(Frequency)) \# Normalize frequencies

\section{Plot allele frequency
distribution}\label{plot-allele-frequency-distribution}

ggplot(allele\_frequencies, aes(x = Allele, y = Proportion, fill =
Group)) + geom\_bar(stat = ``identity'', position = ``dodge'') + \#
Side-by-side bars labs(title = ``Allele Frequency Differences in Maize
and Teosinte'', x = ``Allele'', y = ``Proportion'') +
scale\_fill\_manual(values = c(``Maize'' = ``cornflowerblue'',
``Teosinte'' = ``forestgreen'')) + theme\_minimal()

\end{document}
